Actually Data Checking with the PVC

zsteig opened the pull request.

mjmagee991 committed:
Commented Out Forward Enable Pin

mjmagee991 committed:
Updated and Centralized Pins for PVC V2

mjmagee991 committed:
Updated EVT-core Submodule

zsteig committed:
Testable version of precharging using ADC readings

mjmagee991 committed:
Added Contactor Driver

mjmagee991 committed:
Integrated Contactor Driver

zsteig committed:
ADC reading now works. DOCUMENTATION NEEDED

zsteig committed:
Added pre-op on error logic

mjmagee991 committed:
Fixed and Integrated GFDB Driver

mjmagee991 committed:
Added More Checks to GFDB Driver

mjmagee991 committed:
Updated CAN Interrupt Parameter Struct Name to Match BMS

mjmagee991 committed:
Fixed State Updating During Precharging

zsteig committed:
Changes to prevent disconnected battery allowing precharge

zsteig committed:
Fixed state update messages

zsteig committed:
Fixed state change messages

zsteig committed:
Added heartbeat to dictionary

zsteig committed:
resetting numattemptsmade on success

zsteig committed:
Moved gfdb checking

mjmagee991 committed:
UART added necessary delay for GFD reading to be valid

mjmagee991 committed:
Added CANOpen Heartbeat

mjmagee991 committed:
Debounce GFDB

zsteig committed:
Fix for mis-labeled CAN messages

Commit:
Applied Formatting Changes During GitHub Build

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -1,9 +1,16 @@
 #pragma once
 
+#include "PreCharge/dev/Contactor.hpp"
```
```suggestion
#include <PreCharge/dev/Contactor.hpp>
```

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -378,6 +446,11 @@ class PreCharge {
             .Type = nullptr,
```
Line 410 should be: 
`.Data = (uintptr_t) CO_COBID_TPDO_DEFAULT(0) + NODE_ID,`
This sets the COBID to 18A, which is what it should be, according to the CANopen standard.

mjmagee991 commented on src/PreCharge/PreCharge.cpp:
```diff
@@ -202,35 +275,38 @@ void PreCharge::eStopState() {
-    if (prevState != state) {
-        sendChangePDO();
-    }
     if ((time::millis() - state_start_time) > DISCHARGE_DELAY) {
```
This should be replaced with an ADC check like in the precharge state

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -78,25 +123,25 @@ class PreCharge {
      * @param[in] pc GPIO for precharge contactor
      * @param[in] dc GPIO for discharge contactor
-     * @param[in] cont GPIO for main contactor
+     * @param[in] cont Driver for main contactor
```
```suggestion
     * @param[in] cont Main contactor instance
```

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -78,25 +123,25 @@ class PreCharge {
-              IO::GPIO& eStop, IO::GPIO& pc, IO::GPIO& dc, IO::GPIO& cont,
-              IO::GPIO& apm, IO::GPIO& forward, GFDB::GFDB& gfdb, IO::CAN& can);
+              IO::GPIO& eStop, IO::GPIO& pc, IO::GPIO& dc, Contactor cont,
+              IO::GPIO& apm, GFDB::GFDB& gfdb, IO::CAN& can, MAX22530 MAX);
```
```suggestion
              IO::GPIO& apm, GFDB::GFDB& gfdb, IO::CAN& can, MAX22530 max);
```

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -78,25 +123,25 @@ class PreCharge {
      * The node ID used to identify the device on the CAN network.
      */
-    static constexpr uint8_t NODE_ID = 0x02;
+    static constexpr uint8_t NODE_ID = 0x10;
```
```suggestion
    static constexpr uint8_t NODE_ID = 10;
```

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -107,6 +152,29 @@ class PreCharge {
+     * 
+     * @return int value of PrechargeStatus enum
+    */
+    int getPrechargeStatus();
```
```suggestion
    PreChargeStatus getPrechargeStatus();
```

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -107,6 +152,29 @@ class PreCharge {
+    /**
+     * Get the current expected voltage based on the current precharge time
+     * 
+     * @return float value of expected voltage in Volts
```
```suggestion
     * @return int value of expected voltage in Volts
```

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -236,33 +290,40 @@ class PreCharge {
     IO::GPIO& apm;
     /** GPIO instance to toggle FW_EN_CTL */
-    IO::GPIO& forward;
+    //    IO::GPIO& forward;
```
At this point, I think you can delete all references to forward enable control

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -236,33 +290,40 @@ class PreCharge {
     State state;
     State prevState;
     uint64_t state_start_time;
+    int in_precharge;
```
Make this into an enum. It's hard to tell what it each value means

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -285,6 +346,13 @@ class PreCharge {
         // Sync ID, defaults to 0x80
         {CO_KEY(0x1005, 0, CO_UNSIGNED32 | CO_OBJ_D__R_), nullptr, (uintptr_t) 0x80},
```
```suggestion
        {CO_KEY(0x1005, 0, CO_UNSIGNED32 | CO_OBJ_D__R_), nullptr, (uintptr_t) 0x80,},
```

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -285,6 +346,13 @@ class PreCharge {
 
+        // Enable heartbeat
+        {
+            CO_KEY(0x1017, 0, CO_UNSIGNED16 | CO_OBJ_D__R_),
```
This should have the struct attribute names written explicitly, like `.Key`

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -378,6 +446,11 @@ class PreCharge {
             .Type = nullptr,
             .Data = (uintptr_t) &state,
         },
+        {
```
I think we miscommunicated about this and both added it. You can delete this one. After you do this, be sure to update the object dictionary size

mjmagee991 commented on include/PreCharge/dev/Contactor.hpp:
```diff
@@ -0,0 +1,26 @@
+
+namespace PreCharge {
+
+class Contactor {
```
This file should have some documentation

mjmagee991 committed:
Simplified GFDB requestData()

Commit:
Applied Formatting Changes During GitHub Build

mjmagee991 commented on include/PreCharge/PreCharge.hpp:
```diff
@@ -78,25 +123,25 @@ class PreCharge {
      * Handler running the pre-charge state switching
      */
-    void handle();
+    PVCStatus handle(IO::UART& uart);
```
UART should be removed from this function

mjmagee991 commented on targets/PreCharge/main.cpp:
```diff
@@ -1,28 +1,32 @@
 /**
- * This is the primary State machine handler for the pre-charge board
+ * This is the primary State machine handler for the pre-charge voltage controller (PVC) board
```
```suggestion
 * This is the primary State machine handler for the powertrain voltage controller (PVC) board
```

mjmagee991 commented on targets/PreCharge/main.cpp:
```diff
@@ -100,22 +122,24 @@ int main() {
+        IO::getGPIO<PreCharge::PreCharge::CONT1_PIN>(IO::GPIO::Direction::OUTPUT),
+        IO::getGPIO<PreCharge::PreCharge::CONT2_PIN>(IO::GPIO::Direction::OUTPUT));
+    IO::GPIO& apm = IO::getGPIO<PreCharge::PreCharge::APM_CTL_PIN>(IO::GPIO::Direction::OUTPUT);
+    //    IO::GPIO& forward = IO::getGPIO<IO::Pin::PA_3>(IO::GPIO::Direction::OUTPUT);
```
```suggestion
```

mjmagee991 commented on targets/PreCharge/main.cpp:
```diff
@@ -158,17 +182,24 @@ int main() {
     // Start with everything at 0
     apm.writePin(IO::GPIO::State::LOW);
-    forward.writePin(IO::GPIO::State::LOW);
+    //    forward.writePin(IO::GPIO::State::LOW);
```
```suggestion
```

mjmagee991 commented on targets/manual/main.cpp:
```diff
@@ -39,26 +40,29 @@ void writeGPIO(IO::UART& uart, IO::GPIO& gpio) {
     // Start with everything at 0
     mcuAPMCtl.writePin(IO::GPIO::State::LOW);
-    mcuFWEnCtl.writePin(IO::GPIO::State::LOW);
+    //    mcuFWEnCtl.writePin(IO::GPIO::State::LOW);
```
```suggestion
```

mjmagee991 commented on targets/manual/main.cpp:
```diff
@@ -76,15 +80,15 @@ int main() {
             break;
         case 3:
-            writeGPIO(uart, mcuFWEnCtl);
+            //            writeGPIO(uart, mcuFWEnCtl);
```
This should be removed, and the surrounding code should be changed

mjmagee991 reviewed (changes requested):
A little cleanup needs to be done, but this looks pretty good

mjmagee991 committed:
Fixed Contactor Initialization

mjmagee991 committed:
Added initVolt to Track Initial Voltage Reading

Used to ensure accurate voltage expectations

mjmagee991 committed:
Added voltStatus to Check for Low Voltage

mjmagee991 committed:
Fixed CAN Behavior

Stopped changing into pre-op
Started specifying which node to set to operational and pre-op instead of setting all
Added voltStatus to changePDOMessage
